File: sdks/python/plexspaces/leader_worker.py
# SPDX-License-Identifier: LGPL-2.1-or-later
# Copyright (C) 2025 PlexSpaces Contributors
#
# Leader-worker client for multi-node: same API surface as Rust/TS/Go SDKs.
# Virtual actors are created lazily on first message; use spawn_actor_on_node
# only for non-virtual workers.

from __future__ import annotations

import base64
import json
import re
import urllib.error
import urllib.parse
import urllib.request
from typing import Any, Dict, List, Optional


def _grpc_port_to_http(url: str) -> str:
    """Derive HTTP URL from gRPC node_address (convention: HTTP = gRPC port + 1)."""
    if not url:
        return url
    # Match trailing :port (e.g. :8091 -> :8092)
    match = re.search(r":(\d+)\s*$", url.strip())
    if match:
        port = int(match.group(1)) + 1
        return re.sub(r":\d+\s*$", f":{port}", url.strip())
    return url


class LeaderWorkerClient:
    """
    Client for leader-worker multi-node patterns. Connect to the entry (leader) node
    via HTTP; list worker node IDs and spawn non-virtual actors on specific nodes.
    Virtual actors are created lazily on first message—no explicit spawn or ensure.
    """

    def __init__(self, entry_http_url: str) -> None:
        """
        Args:
            entry_http_url: Base URL of the entry/leader node (e.g. "http://localhost:8092").
        """
        self._entry_url = entry_http_url.rstrip("/")
        self._node_id_to_http: Dict[str, str] = {}

    def list_worker_node_ids(
        self,
        cluster: Optional[str] = None,
        page_size: int = 100,
        page_token: str = "",
    ) -> List[str]:
        """
        List node IDs that can run workers (peers + self). Populates internal
        cache so spawn_actor_on_node can resolve node_id to node HTTP URL.

        Returns:
            List of node_id strings.
        """
        params: Dict[str, str] = {"pageSize": str(page_size)}
        if cluster:
            params["cluster"] = cluster
        if page_token:
            params["pageToken"] = page_token
        qs = urllib.parse.urlencode(params)
        url = f"{self._entry_url}/api/v1/nodes?{qs}"
        req = urllib.request.Request(url, method="GET")
        req.add_header("Accept", "application/json")
        try:
            with urllib.request.urlopen(req, timeout=30) as resp:
                data = json.loads(resp.read().decode())
        except urllib.error.HTTPError as e:
            body = e.read().decode() if e.fp else ""
            raise RuntimeError(f"list_worker_node_ids failed: {e.code} {body}") from e
        except urllib.error.URLError as e:
            raise RuntimeError(f"list_worker_node_ids failed: {e.reason}") from e

        nodes = data.get("nodes") or data.get("nodeRegistrations") or []
        ids: List[str] = []
        for n in nodes:
            node_id = n.get("nodeId") or n.get("node_id") or ""
            node_address = n.get("nodeAddress") or n.get("node_address") or ""
            if node_id:
                ids.append(node_id)
                if node_address:
                    self._node_id_to_http[node_id] = _grpc_port_to_http(node_address)
        return ids

    def spawn_actor_on_node(
        self,
        node_id: str,
        actor_type: str,
        actor_id: str = "",
        initial_state: bytes = b"",
        config: Optional[Dict[str, Any]] = None,
        labels: Optional[Dict[str, str]] = None,
    ) -> str:
        """
        Spawn a non-virtual actor on a specific node. The node must be reachable;
        call list_worker_node_ids first so the client can resolve node_id to URL.

        Returns:
            Actor ref string (e.g. "worker-ulid@node_id").
        """
        node_http = self._node_id_to_http.get(node_id)
        if not node_http:
            raise RuntimeError(
                f"unknown node_id {node_id!r}; call list_worker_node_ids first"
            )
        payload = {"actorType": actor_type}
        if actor_id:
            payload["actorId"] = actor_id
        if initial_state:
            payload["initialState"] = base64.b64encode(initial_state).decode()
        if config is not None:
            payload["config"] = config
        if labels is not None:
            payload["labels"] = labels
        body = json.dumps(payload).encode()
        url = f"{node_http.rstrip('/')}/api/v1/actors/spawn"
        req = urllib.request.Request(url, data=body, method="POST")
        req.add_header("Content-Type", "application/json")
        req.add_header("Accept", "application/json")
        try:
            with urllib.request.urlopen(req, timeout=30) as resp:
                out = json.loads(resp.read().decode())
        except urllib.error.HTTPError as e:
            body_str = e.read().decode() if e.fp else ""
            raise RuntimeError(f"spawn_actor_on_node failed: {e.code} {body_str}") from e
        except urllib.error.URLError as e:
            raise RuntimeError(f"spawn_actor_on_node failed: {e.reason}") from e

        actor_ref = out.get("actorRef") or out.get("actor_ref") or ""
        if not actor_ref:
            raise RuntimeError("spawn_actor_on_node returned empty actorRef")
        return actor_ref


def list_worker_node_ids(
    entry_http_url: str,
    cluster: Optional[str] = None,
    page_size: int = 100,
) -> List[str]:
    """
    Convenience: list worker node IDs using a one-off client.
    For multiple calls or spawn_actor_on_node, use LeaderWorkerClient.
    """
    client = LeaderWorkerClient(entry_http_url)
    return client.list_worker_node_ids(cluster=cluster, page_size=page_size)

File: sdks/go/plexspaces/leader_worker.go
// SPDX-License-Identifier: LGPL-2.1-or-later
// Copyright (C) 2025 PlexSpaces Contributors
//
// Leader-worker client for multi-node: same API surface as Rust/Python/TypeScript SDKs.
// Virtual actors are created lazily on first message; use SpawnActorOnNode
// only for non-virtual workers.
//
//go:build !wasm

package plexspaces

import (
	"bytes"
	"encoding/base64"
	"encoding/json"
	"fmt"
	"io"
	"net/http"
	"net/url"
	"regexp"
	"strconv"
	"strings"
)

// LeaderWorkerClient is the client for leader-worker multi-node patterns.
// Connect to the entry (leader) node via HTTP; list worker node IDs and spawn
// non-virtual actors on specific nodes. Virtual actors are created lazily on
// first message—no explicit spawn or ensure.
type LeaderWorkerClient struct {
	entryURL      string
	nodeIDToHTTP  map[string]string
}

// NewLeaderWorkerClient creates a client for the entry/leader node.
// entryHTTPURL is the base URL of the entry node (e.g. "http://localhost:8092").
func NewLeaderWorkerClient(entryHTTPURL string) *LeaderWorkerClient {
	entryURL := strings.TrimSuffix(entryHTTPURL, "/")
	return &LeaderWorkerClient{
		entryURL:     entryURL,
		nodeIDToHTTP: make(map[string]string),
	}
}

var grpcPortRE = regexp.MustCompile(`:(\d+)\s*$`)

func grpcPortToHTTP(addr string) string {
	if addr == "" {
		return addr
	}
	addr = strings.TrimSpace(addr)
	sub := grpcPortRE.FindStringSubmatch(addr)
	if len(sub) != 2 {
		return addr
	}
	port, _ := strconv.Atoi(sub[1])
	port++
	return grpcPortRE.ReplaceAllString(addr, fmt.Sprintf(":%d", port))
}

// nodeRegistration is the JSON shape for list nodes response (camelCase or snake_case).
type nodeRegistration struct {
	NodeID      string `json:"nodeId"`
	NodeIDSnake string `json:"node_id"`
	NodeAddress string `json:"nodeAddress"`
	NodeAddressSnake string `json:"node_address"`
}

// listNodesResponse is the JSON shape for GET /api/v1/nodes.
type listNodesResponse struct {
	Nodes              []nodeRegistration `json:"nodes"`
	NodeRegistrations  []nodeRegistration `json:"nodeRegistrations"`
}

// ListWorkerNodeIds lists node IDs that can run workers (peers + self).
// It populates the internal cache so SpawnActorOnNode can resolve nodeID to HTTP URL.
// cluster is optional; pageSize defaults to 100.
func (c *LeaderWorkerClient) ListWorkerNodeIds(cluster string, pageSize int, pageToken string) ([]string, error) {
	if pageSize <= 0 {
		pageSize = 100
	}
	params := url.Values{}
	params.Set("pageSize", fmt.Sprintf("%d", pageSize))
	if cluster != "" {
		params.Set("cluster", cluster)
	}
	if pageToken != "" {
		params.Set("pageToken", pageToken)
	}
	reqURL := c.entryURL + "/api/v1/nodes?" + params.Encode()
	req, err := http.NewRequest(http.MethodGet, reqURL, nil)
	if err != nil {
		return nil, err
	}
	req.Header.Set("Accept", "application/json")
	resp, err := http.DefaultClient.Do(req)
	if err != nil {
		return nil, err
	}
	defer resp.Body.Close()
	if resp.StatusCode != http.StatusOK {
		body, _ := io.ReadAll(resp.Body)
		return nil, fmt.Errorf("list_worker_node_ids failed: %d %s", resp.StatusCode, string(body))
	}
	var data listNodesResponse
	if err := json.NewDecoder(resp.Body).Decode(&data); err != nil {
		return nil, err
	}
	nodes := data.Nodes
	if len(nodes) == 0 {
		nodes = data.NodeRegistrations
	}
	ids := make([]string, 0, len(nodes))
	for _, n := range nodes {
		nodeID := n.NodeID
		if nodeID == "" {
			nodeID = n.NodeIDSnake
		}
		if nodeID == "" {
			continue
		}
		ids = append(ids, nodeID)
		addr := n.NodeAddress
		if addr == "" {
			addr = n.NodeAddressSnake
		}
		if addr != "" {
			c.nodeIDToHTTP[nodeID] = grpcPortToHTTP(addr)
		}
	}
	return ids, nil
}

// SpawnActorRequest is the JSON body for POST /api/v1/actors/spawn.
type spawnActorRequest struct {
	ActorType     string            `json:"actorType"`
	ActorID       string            `json:"actorId,omitempty"`
	InitialState  string            `json:"initialState,omitempty"` // base64
	Config        map[string]any    `json:"config,omitempty"`
	Labels        map[string]string `json:"labels,omitempty"`
}

// SpawnActorResponse is the JSON response from spawn.
type spawnActorResponse struct {
	ActorRef     string `json:"actorRef"`
	ActorRefSnake string `json:"actor_ref"`
}

// SpawnActorOnNode spawns a non-virtual actor on the given node.
// The node must be known; call ListWorkerNodeIds first.
// Returns the actor ref (e.g. "worker-ulid@node_id").
func (c *LeaderWorkerClient) SpawnActorOnNode(
	nodeID string,
	actorType string,
	actorID string,
	initialState []byte,
	config map[string]any,
	labels map[string]string,
) (string, error) {
	nodeHTTP, ok := c.nodeIDToHTTP[nodeID]
	if !ok {
		return "", fmt.Errorf("unknown node_id %q; call ListWorkerNodeIds first", nodeID)
	}
	reqBody := spawnActorRequest{ActorType: actorType, ActorID: actorID}
	if len(initialState) > 0 {
		reqBody.InitialState = base64.StdEncoding.EncodeToString(initialState)
	}
	if config != nil {
		reqBody.Config = config
	}
	if labels != nil {
		reqBody.Labels = labels
	}
	body, err := json.Marshal(reqBody)
	if err != nil {
		return "", err
	}
	reqURL := strings.TrimSuffix(nodeHTTP, "/") + "/api/v1/actors/spawn"
	req, err := http.NewRequest(http.MethodPost, reqURL, bytes.NewReader(body))
	if err != nil {
		return "", err
	}
	req.Header.Set("Content-Type", "application/json")
	req.Header.Set("Accept", "application/json")
	resp, err := http.DefaultClient.Do(req)
	if err != nil {
		return "", err
	}
	defer resp.Body.Close()
	if resp.StatusCode != http.StatusOK {
		respBody, _ := io.ReadAll(resp.Body)
		return "", fmt.Errorf("spawn_actor_on_node failed: %d %s", resp.StatusCode, string(respBody))
	}
	var out spawnActorResponse
	if err := json.NewDecoder(resp.Body).Decode(&out); err != nil {
		return "", err
	}
	actorRef := out.ActorRef
	if actorRef == "" {
		actorRef = out.ActorRefSnake
	}
	if actorRef == "" {
		return "", fmt.Errorf("spawn_actor_on_node returned empty actorRef")
	}
	return actorRef, nil
}

// ListWorkerNodeIds is a convenience that lists worker node IDs using a one-off client.
// For multiple calls or SpawnActorOnNode, use LeaderWorkerClient.
func ListWorkerNodeIds(entryHTTPURL string, cluster string, pageSize int) ([]string, error) {
	client := NewLeaderWorkerClient(entryHTTPURL)
	return client.ListWorkerNodeIds(cluster, pageSize, "")
}
